Objects/dictobject.c

```diff
@@ -2591,6 +2581,102 @@ dict_merge(PyObject *a, PyObject *b, int override)
     return 0;
 }
 
+/* Implementation of d.update(obj). There are two cases:
+ * - if obj has a "keys" method, assume that it is a mapping
+ * - otherwise, obj must be an iterable of (key, value) pairs
+ */
+static int
+dict_merge_from_any(PyObject *self, PyObject *obj)
+{
+    /* Fast path for dict subclasses */
+    if (PyDict_Check(obj)) {
+        return dict_merge(self, obj, 1);
+    }
+
+    _Py_IDENTIFIER(keys);
+    PyObject *func;
+    if (_PyObject_LookupAttrId(obj, &PyId_keys, &func) < 0) {
+        return -1;
+    }
+    if (func != NULL) {
+        Py_DECREF(func);
+        return dict_merge(self, obj, 1);
+    }
+    else {
+        return PyDict_MergeFromSeq2(self, obj, 1);
+    }
+}
+
+/* Insert the key:value pairs from the arrays keys and values (of length n).
+ * If a key is not a string, raise TypeError */
+static int
+dict_merge_from_kw_arrays(PyDictObject *mp, PyObject *const *keys, PyObject *const *values, Py_ssize_t n)
+{
+    /* Do one big resize at the start, rather than
+     * incrementally resizing as we insert new items.  Expect
+     * that there will be no (or few) overlapping keys.
+     */
+    if (USABLE_FRACTION(mp->ma_keys->dk_size) < n) {
+        if (dictresize(mp, ESTIMATE_SIZE(mp->ma_used + n))) {
+           return -1;
+        }
+    }
+
+    for (Py_ssize_t i = 0; i < n; i++) {
+        PyObject *key = keys[i];
```

Comment:
Would you mind to add `PyObject *value = values[i];` to better show that we iterate on zip(keys, values)?